cpp/src/arrow/array/builder_base.h

```diff
@@ -177,6 +178,15 @@ class ARROW_EXPORT ArrayBuilder {
 
   static Status TrimBuffer(const int64_t bytes_filled, ResizableBuffer* buffer);
 
+  /// \brief Finish to an array of the specified ArrayType
+  template <typename ArrayType>
+  Status FinishTyped(std::shared_ptr<ArrayType>* out) {
+    std::shared_ptr<Array> out_untyped;
+    ARROW_RETURN_NOT_OK(Finish(&out_untyped));
+    *out = std::static_pointer_cast<ArrayType>(std::move(out_untyped));
+    return Status::OK();
+  }
```

Comment:
One option is to Call this `Finish` simply and add a DCHECK to assert that the runtime type matches `ArrayType`. I tried this here:

https://github.com/wesm/arrow/commit/0be7cdab8730c2bfc23d6b9cf075f12dd0fb2176

I'm not sure what pitfalls there might be from that

Reply:
That's an option. I avoided it because it seems sloppy/user-hostile to allow something to compile which could easily be caught as an error:

```
Int64Builder i64_b;
// ...
std::shared_ptr<Int32Array> i32_a;
i64_b.Finish(&i32_a); // would compile, but fail at runtime
```

Reply:
Another way would be to make this a non-member method:

```cpp
Int64Builder builder;
// ...
std::shared_ptr<Int64Array> array;
Finish(&builder, &array); // type match guaranteed at compile time
```

Reply:
Gotcha. Well, I think it's OK to accept things as they are now, the boilerplate does not seem too onerous